rocAL/source/readers/image/numpy_data_reader.cpp

```diff
@@ -0,0 +1,642 @@
+    if (it == _header_cache.end()) {
+        return false;
+    } else {
+        header = it->second;
+        return true;
+    }
+}
+
+void NumpyDataReader::update_header_cache(const std::string& file_name, const NumpyHeaderData& value) {
+    std::unique_lock<std::mutex> cache_lock(_cache_mutex);
+    _header_cache[file_name] = value;
+}
+
+const RocalTensorDataType NumpyDataReader::get_dtype(const std::string& format) {
+    if (format == "u1") return RocalTensorDataType::UINT8;
+    if (format == "u2") THROW("uint16_t dtype not supported in rocAL");
+    if (format == "u4") return RocalTensorDataType::UINT32;
+    if (format == "u8") THROW("uint64_t dtype not supported in rocAL");
+    if (format == "i1") return RocalTensorDataType::INT8;
+    if (format == "i2") return RocalTensorDataType::INT16;
+    if (format == "i4") return RocalTensorDataType::INT32;
+    if (format == "i8") THROW("int64_t dtype not supported in rocAL");
+    if (format == "f2")
+#if defined(AMD_FP16_SUPPORT)
+        return RocalTensorDataType::FP16;
+#else
+        THROW("FLOAT16 type tensor not supported")
+#endif
+    if (format == "f4") return RocalTensorDataType::FP32;
+    if (format == "f8") THROW("double dtype not supported in rocAL");
+    THROW("Unknown Numpy dtype string");
+}
+
+template <size_t N>
+void NumpyDataReader::skip_char(const char*& ptr, const char (&what)[N]) {
+    if (strncmp(ptr, what, N - 1)) {
+        ERR("Found wrong symbol during parsing, expected symbol: " + std::string(what));
+        _header_parsing_failed = true;
+        return;
+    }
+    ptr += N - 1;
+}
+
+template <size_t N>
+bool NumpyDataReader::try_skip_char(const char*& ptr, const char (&what)[N]) {
+    if (!strncmp(ptr, what, N - 1)) {
+        ptr += N - 1;
+        return true;
+    } else {
+        return false;
+    }
+}
+
+template <size_t N>
+void NumpyDataReader::skip_field(const char*& ptr, const char (&name)[N]) {
+    while (std::isspace(*ptr)) ptr++;
+    CALL_AND_CHECK_FLAG(skip_char(ptr, "'"));
+    CALL_AND_CHECK_FLAG(skip_char(ptr, name));
+    CALL_AND_CHECK_FLAG(skip_char(ptr, "'"));
+    while (std::isspace(*ptr)) ptr++;
+    CALL_AND_CHECK_FLAG(skip_char(ptr, ":"));
+    while (std::isspace(*ptr)) ptr++;
+}
+
+template <typename T>
+T NumpyDataReader::parse_int(const char*& ptr) {
+    char* out_ptr = const_cast<char*>(ptr);  // strtol takes a non-const pointer
+    T value = static_cast<T>(strtol(ptr, &out_ptr, 10));
+    if (out_ptr == ptr) {
+        ERR("Parse error: expected a number.");
+        _header_parsing_failed = true;
+        return value;
+    }
+    ptr = out_ptr;
+    return value;
+}
+
+std::string NumpyDataReader::parse_string(const char*& input, char delim_start, char delim_end) {
+    if (*input++ != delim_start) {
+        ERR("Expected \'" + std::to_string(delim_start) + "\'");
+        _header_parsing_failed = true;
+        return "";
+    }
+    std::string out;
+    for (; *input != '\0'; input++) {
```

Comment:
please get rid of this switch statement and use if with ||.
if (input[1] == '\\' || ....) {
    out += input[1];
}